test/Microsoft.ML.TestFramework/DataPipe/TestDataPipeBase.cs

```diff
@@ -151,14 +151,18 @@ public abstract partial class TestDataPipeBase : TestDataViewBase
         private void CheckSameSchemaShape(SchemaShape promised, SchemaShape delivered)
         {
             Assert.True(promised.Count == delivered.Count);
-            var sortedCols1 = promised.OrderBy(x => x.Name);
-            var sortedCols2 = delivered.OrderBy(x => x.Name);
+            var promisedCols = promised.OrderBy(x => x.Name);
+            var deliveredCols = delivered.OrderBy(x => x.Name);
 
-            foreach (var (x, y) in sortedCols1.Zip(sortedCols2, (x, y) => (x, y)))
+            foreach (var (p, d) in promisedCols.Zip(deliveredCols, (p, d) => (p, d)))
             {
-                Assert.Equal(x.Name, y.Name);
+                Assert.Equal(p.Name, d.Name);
                 // We want the 'promised' metadata to be a superset of 'delivered'.
-                Assert.True(y.IsCompatibleWith(x), $"Mismatch on {x.Name}");
+                Assert.True(d.IsCompatibleWith(p), $"Mismatch on {p.Name}, there was a mismatch, or some unexpected annotations was present.");
```

Comment:
>Assert.True(d.IsCompatibleWith(p), $"Mismatch on {p.Name}, there was a mismatch, or some unexpected annotations was present."); [](start = 16, length = 127)

I'm trying to come up with some example where this wouldn't be true....
I think previous statement was made based on fact what  what `SchemaShape` has less information than `Schema` and we don't have actual data, so in some cases we will add annotations which was impossible to predict during GetOutputSchema time.

But nothing comes to my mind, and since tests are not failing, I would guess this is our current state of annotation business.

Reply:
> so in some cases we will add annotations which was impossible to predict during GetOutputSchema time.

Maybe. The way the test worked, it made sure the promise was a superset of what was delivered -- that is, the estimator's annotation list would be a sort of "semi-promise" like, "ok, this annotation may be there or not, but if it *is* there, it will have this type," and so on. At least, so I must interpret this test as previously written.

There is one case in which this could potentially have been valuable, the key-to-vector annotation, as discussed in depth in issue #3380 (the special case where it was a known size vector that was of length *precisely one*). But that was such an obscure situation and such a micro-optimization that I hardly felt it worthwhile.

And anyway, the `IEstimator` did things in the wrong way, it only said the slot names would be there if it was scalar, whereas if it actually wanted to capture this "possibility" it should have said that the slot names *could* be there if scalar or vector (but not var-vector!). But if you're going to go *that* far, I'd argue that the usefulness of `SchemaShape` annotations data is utterly useless.

I could have imagined a "richer" schema shape where there were differing degrees of certainty -- "this will definitely be here, this *might* be here," etc., but in the end I decided with saying, "you know what, let's just make it a simple promise."

One nice thing is that `IEstimator` is *still* not possible to fully implement, so hypothetical future authors might be able to refine this.